server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1356,7 +1355,7 @@ private static ImmutableOpenMap<ShardId, ShardSnapshotStatus> processWaitingShar
             }
         }
         if (snapshotChanged) {
-            return shards.build();
+            return shards;
```

Comment:
We need to wrap into unmodifiable collection here.

Reply:
Possibly...this is a private method, and the map only goes to the `SnapshotsInProgress.Entry` class. However, that constructor doesn't call `Map.copyOf` on thie collection, whereas it does on other arguments. Looks like these classes needs a lot more consistency in its use of maps & immutability ownership now. What are your thoughts?

Reply:
Yeah I think we need to ensure there's consistency on how we handle the mutable maps on all of these touch points. The classes that used to receive ImmutableOpenMap before should get an immutable map always, or a copy if needs to happen. As you mentioned, it might be better to do this in the constructors if it's not done already.

We need to prevent hard to track bug where a map is modified after it has been given to another consumer, so perhaps a pattern to follow is something like:

```
var map = new HashMap<...>();

... do something with the map

map = Collections.unmodifiableMap(map) // where we would've normally done ImmutableOpenMap.build().

```

Reply:
I've had a general look at maps in this PR:
- Maps in fields are immutable through `Map.copyOf`
- Maps returned from methods are shallow immutable through `unmodifiableMap`
- Maps that are passed into a method & then go out of scope after don't matter - its up to the method called to copy appropriately if it needs to